test/extension/handlers/topologymutation/handler_test.go

```diff
@@ -0,0 +1,406 @@
+				},
+			},
+		},
+		{
+			name: "Error with no variables found for the dockerCluster patch",
+			requestItems: []runtimehooksv1.GeneratePatchesRequestItem{
+				requestItem("1", dockerClusterTemplate, emptyVars),
+			},
+			expectedResponse: &runtimehooksv1.GeneratePatchesResponse{
+				CommonResponse: runtimehooksv1.CommonResponse{
+					Status: runtimehooksv1.ResponseStatusFailure,
+				},
+			},
+		},
+		{
+			name: "Error with no variables found for the kubeadmControlPlane patch",
+			requestItems: []runtimehooksv1.GeneratePatchesRequestItem{
+				requestItem("1", kubeadmControlPlane, emptyVars),
+			},
+			expectedResponse: &runtimehooksv1.GeneratePatchesResponse{
+				CommonResponse: runtimehooksv1.CommonResponse{
+					Status: runtimehooksv1.ResponseStatusFailure,
+				},
+			},
+		},
+		{
+			name: "Error with no variables found for the kubeadmConfigTemplate patch",
+			requestItems: []runtimehooksv1.GeneratePatchesRequestItem{
+				requestItem("1", kubeadmConfigTemplate, emptyVars),
+			},
+			expectedResponse: &runtimehooksv1.GeneratePatchesResponse{
+				CommonResponse: runtimehooksv1.CommonResponse{
+					Status: runtimehooksv1.ResponseStatusFailure,
+				},
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			response := &runtimehooksv1.GeneratePatchesResponse{}
+			request := &runtimehooksv1.GeneratePatchesRequest{Items: tt.requestItems}
+			h.GeneratePatches(context.Background(), request, response)
+
+			// Expect all response fields to be as expected. responseItems are ignored here and tested below.
+			// Ignore the message to not compare error strings.
+			g.Expect(tt.expectedResponse).To(EqualObject(response, IgnorePaths{{"items"}, {"message"}}))
+			// For each item in the response check that the patches are the same.
+			// Note that the order of the individual patch operations in Items[].Patch is not determinate so we unmarshal
+			// to an array and check that the arrays hold equivalent items.
+			for i, item := range response.Items {
+				expectedItem := tt.expectedResponse.Items[i]
+				g.Expect(item.PatchType).To(Equal(expectedItem.PatchType))
+				g.Expect(item.UID).To(Equal(expectedItem.UID))
+
+				var actualPatchOps []map[string]interface{}
+				var expectedPatchOps []map[string]interface{}
+				g.Expect(json.Unmarshal(item.Patch, &actualPatchOps)).To(Succeed())
+				g.Expect(json.Unmarshal(expectedItem.Patch, &expectedPatchOps)).To(Succeed())
+				g.Expect(actualPatchOps).To(ConsistOf(expectedPatchOps))
+			}
+		})
+	}
+}
+
+// toJSONCompact is used to be able to write JSON values in a readable manner.
+func toJSONCompact(value string) []byte {
+	var compactValue bytes.Buffer
+	if err := json.Compact(&compactValue, []byte(value)); err != nil {
+		panic(err)
+	}
+	return compactValue.Bytes()
+}
+
+// objectToJson marchals the object and returns a byte array. This function panics on any error.
+func objectToJSON(val interface{}) []byte {
```

Comment:
```suggestion
func toJSON(val interface{}) []byte {
```
shorter and closer to reality. We convert everythign to JSON (and consistent with toJSONCompact)